guitar-api.yaml

```diff
@@ -0,0 +1,122 @@
+openapi: 3.0.3
+info:
+  title: Guitar API
+  description: |-
+   Reference Basic infromation on guitar models and the musicians that have used them at various points in their careers.
+    - [Guitar API Repository](https://github.com/Bobs-code/Guitar-API)
+  contact:
+    email: robertgriffintm@gmail.com
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
+  version: 1.0.0
+servers:
+  - url: http://localhost:8080
+tags:
+  - name: guitar
+    description: Everything about popular guitar models
+    externalDocs:
+      description: Find out more
+      url: http://swagger.io
+paths:
+  /guitar:
```

Comment:
**Strong suggestion**: the GET (and only the GET) needs to live at `/guitar/{guitar_id}` rather than `/guitar`. The reason is the difference between **path** vs **query** parameters:

- **path**: `/guitar/123`
- **query**: `/guitar?guitar_id=123`

There’s nothing wrong with the way you’ve done it! It’s perfectly valid. But it’s sort of a common convention to **prefer path parameters when possible for required parameters**. And whenever you have a path parameter, it must go in the path (URL).

The reason is that REST philosophy tries to structure URLs in a way that go from a parent (generic) resource -> children (specific) resources. Query params break that pattern because they **always must be stuck at the end**. So when you get into longer URLs, they not only break that left-to-right order; when you get a lot of query params it just gets messy:

```
/some/really/really/long/url/?guitar_id=123&something_else=foo&another_thing=bar
```

That said, I _have_ used APIs that have required query params. You’re allowed to do that. But I have never encountered a _well-designed_ API that has done this; generally query params are for optional refinement (which you don’t have to worry about now at this stage of your API, but if an opportunity pops up I notice I’ll point it out)

Reply:
Ah sorry; let me revise that: DELETE and PUT also need to live at `/guitar/{guitar_id}` too because they also need that param.

`POST` (create) doesn’t, because generally-speaking, we **never** want the user to be able to pick their ID (for security and safety reasons); we will assign one for them. So because they can’t provide an ID, they won’t.

Reply:
> But it’s sort of a common convention to prefer path parameters when possible for required parameters. And whenever you have a path parameter, it must go in the path (URL).

That makes sense. I think the issue I have is that in the Go Std Lib, there seems to be little support for setting up clean path parameters. The net/http package seems to only be able to grab URL query parameters. There are quite a few different third party/community packages that handle this and allow for simple path params to be set up. Of the many examples I've seen of folks setting up routers for their API's, all of them opt for [gorilla/mux](https://github.com/gorilla/mux) or [chi](https://github.com/go-chi/chi) instead of the standard net/http package. 

I think I may have mentioned this elsewhere so apologies for the repetition, but I have been thinking about using one of these down the road for this project or another. While Gorilla seems more straightforward, it is no longer supported as of Dec '22. Knowing how quickly things evolve in tech, I'm sure it will serve its purpose for some time to come, but I am not counting on it remaining reliable over the future. Chi seems to be the way forward.

Reply:
Ah gotcha. I am not writing Go enough to know all the ins-and-outs of the state of the ecosystem at the moment. But you seem to be figuring it out!

From an outside perspective, you seem to be in the right headspace. I see you asking questions like:

- What does the Go std library provide? What does it make hard/easy to do?
- What is the community doing right now? Where is there consensus? Where is there debate?

This is all center-of-wheelhouse for an engineer, as someone whose job it is to take a requirement (path params) and figure out how to make it work in their tech stack. And smoothing over the parts that aren’t a solved problem. I know there’s a lot of ambiguity here, but that’s also why there’s a lot of money here as well :)

Reply:
Also though you are more up-to-date than me with Go, I do know that path parameters are foundational parts of APIs, so I do know that nearly _every_ Go developer is dealing with that problem.

I also feel pretty confident that Go wouldn’t prevent you from using path parameters, as that’s core REST behavior. And as Go is built on REST principles and is an internet-aware language used to build APIs, it may just be a matter of finding the right solution there.